ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/HealthcheckResource.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.resources;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.rest.entity.HealthcheckResponse;
+import io.confluent.ksql.rest.entity.Versions;
+import io.confluent.ksql.rest.healthcheck.HealthcheckAgent;
+import io.confluent.ksql.rest.server.KsqlRestConfig;
+import io.confluent.ksql.services.ServiceContext;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.function.Supplier;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+@Path("/healthcheck")
+@Produces({Versions.KSQL_V1_JSON, MediaType.APPLICATION_JSON})
+public class HealthcheckResource {
+  private final HealthcheckAgent healthcheckAgent;
+  private final Duration healthcheckInterval;
+  private final Supplier<Long> currentTimeSupplier;
+  private final ResponseCache responseCache;
+
+  @VisibleForTesting
+  HealthcheckResource(
+      final HealthcheckAgent healthcheckAgent,
+      final Duration healthcheckInterval,
+      final Supplier<Long> currentTimeSupplier
+  ) {
+    this.healthcheckAgent = Objects.requireNonNull(healthcheckAgent, "healthcheckAgent");
+    this.healthcheckInterval = Objects.requireNonNull(healthcheckInterval, "healthcheckInterval");
+    this.currentTimeSupplier = Objects.requireNonNull(currentTimeSupplier, "currentTimeSupplier");
+    this.responseCache = new ResponseCache(currentTimeSupplier);
+  }
+
+  @GET
+  public Response checkHealth() {
+    return Response.ok(getResponse()).build();
+  }
+
+  private HealthcheckResponse getResponse() {
+    if (responseCache.isEmpty() || timeSinceLastResponse().compareTo(healthcheckInterval) > 0) {
```

Comment:
it would be cleaner to move the expiry into the cache, e.g.:

```
final Optional<HealthCheckResponse> response = responseCache.get();
if (response.isPresent()) {
    return response.get();
} 
final HealthcheckResponse fresh = healtcheckAgent.checkHealth();
responseCache.cache(fresh);
return fresh;
```

and in the Cache, get() can return empty if nothing is cached or if its expired.

Reply:
Done, though I wonder if it's clear from just reading 
```
final Optional<HealthCheckResponse> response = responseCache.get();
if (response.isPresent()) {
    return response.get();
} 
```
that the cache handles expiration (rather than returning any previously cached response). Maybe I should update the cache name to be `ExpiringResponseCache` or something similar?

Reply:
Maybe if we change the get() method to getIfPresent() it'll be clearer?

Reply:
I think the fact that the return type is an `Optional` already indicates that "if present" part. It was the fact that the cache expires that I didn't think was captured. I don't want to block the PR on this though, can always open a follow-up to change the name.

Reply:
getIfUnexpired() then to indicate expiration behavior? I do agree this is a minor point though.